### PR TITLE
dynamic-client-authentication unregister function

### PR DESCRIPTION
**Pull request by milanperera**

### CODE REVIEW

File: components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/RegistrationService.java [hunk 1]
```diff
@@ -31,4 +32,7 @@
     @POST
     Response register(RegistrationProfile profile);
 
+    @DELETE
+    Response unregister(UnregistrationProfile profile);
```
Comment (prabathabey) [inline components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/RegistrationService.java:36]: Add an appropriate comments to the interface methods.

File: components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java [hunk 2]
```diff
@@ -75,6 +77,22 @@ public Response register(RegistrationProfile profile) {
...
     }
 
+    @DELETE
+    @Override
+    public Response unregister(UnregistrationProfile profile) {
+        String applicationName = profile.getApplicationName();
+        String consumerKey = profile.getConsumerKey();
+        String userId = profile.getUserId();
+        try {
+            this.unregisterApplication(userId, applicationName, consumerKey);
```
Comment (DilanUA) [inline components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java:87]: this. is unnecessary here to call unregisterApplication()

File: components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java [hunk 3]
```diff
@@ -232,4 +250,45 @@ public OAuthApplicationInfo createOAuthApplication(
         return null;
     }
 
+    public void unregisterApplication(String userId, String applicationName, String consumerKey)
```
Comment (DilanUA) [inline components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java:253]: shouldn't this be private as of registerApplication()

File: components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java [hunk 4]
```diff
@@ -232,4 +250,45 @@ public OAuthApplicationInfo createOAuthApplication(
...
+        PrivilegedCarbonContext.startTenantFlow();
+        PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantDomain(tenantDomain, true);
+        PrivilegedCarbonContext.getThreadLocalCarbonContext().setUsername(userName);
+
+        if (userId == null || userId.isEmpty()) {
+            throw new APIManagementException("Error occurred while unregistering Application: userId cannot be null/empty");
+        }
+        try {
+            OAuthAdminService oAuthAdminService = new OAuthAdminService();
+            OAuthConsumerAppDTO oAuthConsumerAppDTO = oAuthAdminService.getOAuthApplicationData(consumerKey);
```
Comment (DilanUA) [inline components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java:269]: what is  the usage of oAuthConsumerAppDTO here? don't see any usage here.

File: components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java [hunk 5]
```diff
@@ -232,4 +250,45 @@ public OAuthApplicationInfo createOAuthApplication(
...
+
+            if (oAuthConsumerAppDTO == null) {
+                throw new APIManagementException("Couldn't retrieve OAuth Consumer Application associated with the " +
+                                                 "given consumer key: " + consumerKey);
+            }
+            oAuthAdminService.removeOAuthApplicationData(consumerKey);
+
+            ApplicationManagementService appMgtService = ApplicationManagementService.getInstance();
```
Comment (DilanUA) [inline components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java:277]: shouldn't here be a null check for appMgtService?

File: components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java [hunk 6]
```diff
@@ -232,4 +250,45 @@ public OAuthApplicationInfo createOAuthApplication(
...
+            }
+            appMgtService.deleteApplication(applicationName);
+
+        } catch (IdentityApplicationManagementException e) {
+            APIUtil.handleException("Error occurred while removing ServiceProvider for app " + applicationName, e);
+        } catch (Exception e) {
+            APIUtil.handleException("Error occurred while removing OAuthApp " + applicationName, e);
+        } finally {
+            PrivilegedCarbonContext.endTenantFlow();
+            PrivilegedCarbonContext.getThreadLocalCarbonContext().setUsername(baseUser);
```
Comment (DilanUA) [inline components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java:291]: what's the usage of setting username after ending tenant flow?

File: components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java (hunk 4 above)
Comment (milanperera) [inline components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java:269]: It is used to check whether the application is already there for given consumer key.

File: components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java (hunk 2 above)
Comment (milanperera) [inline components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java:87]: This class is removed and the function is introduced in commit [1].
[1] https://github.com/wso2/carbon-device-mgt/commit/8073d46f58c8eede4d1960b026cc32b7b23b28cb

File: components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java (hunk 3 above)
Comment (milanperera) [inline components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java:253]: This class is removed and the function is introduced as an util method in DynamicClientRegistrationUtil class in commit [1].
[1] https://github.com/wso2/carbon-device-mgt/commit/8073d46f58c8eede4d1960b026cc32b7b23b28cb

File: components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java (hunk 5 above)
Comment (milanperera) [inline components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java:277]: Fixed in PR [1].
[1] https://github.com/wso2/carbon-device-mgt/pull/27

File: components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/RegistrationService.java (hunk 1 above)
Comment (milanperera) [inline components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/RegistrationService.java:36]: Added in PR [1].
[1] https://github.com/wso2/carbon-device-mgt/pull/27

File: components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java (hunk 6 above)
Comment (milanperera) [inline components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java:291]: When we registering an Oauth application, we create them under specific tenant for a given application owner. This owner is only known by this tenant. Therefor before we do anything specific to this tenant, we keep the current thread local user as a base user. Then we start tenant flow and set the application owner as the thread local user in order to use for the tenant specific logic. Finally, after ending the tenant flow, we set back the base user as the thread local user.

File: components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/UnregistrationProfile.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
+/*
+* Copyright (c) 2014, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (prabathabey) [inline components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/UnregistrationProfile.java:2]: License header used is invalid. Please fix.

Comment (milanperera) [inline components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/UnregistrationProfile.java:2]: This class had been removed and introduced query params instead.

**Pull request merged**